### PR TITLE
[dnssd-server] add `otDnssdGetCounters` for telemetry

### PR DESCRIPTION
**Pull request by superwhd**
This PR implements `otDnssdGetCounters` to report the statistics of the DNS-SD server.

### CODE REVIEW

File: src/core/net/dnssd_server.hpp [hunk 1]
```diff
@@ -300,6 +315,7 @@ class Server : public InstanceLocator, private NonCopyable
         NameCompressInfo mCompressInfo;
         Ip6::MessageInfo mMessageInfo;
         TimeMilli        mStartTime;
+        Server *         mServer;
```
Comment (simonlingoogle) [inline src/core/net/dnssd_server.hpp:318]: `Server` is an `InstanceLocator`. 

We can get `Server` directly if `QueryTransaction` is also a `InstanceLocator`?

Comment (superwhd) [reply src/core/net/dnssd_server.hpp:318]: If we use `InstanceLocator` here, we have to provide an `Instance &` to the constructor of `InstanceLocator`. That means we may need to pass the `Instance &` as an argument to the constructor of `QueryTransaction`. 

However, `Server` owns an array of `QueryTransaction`s . It's tricky to construct such an array of objects who don't have a default constructor.

Comment (simonlingoogle) [reply src/core/net/dnssd_server.hpp:318]: Good point. Thanks.

**Review by simonlingoogle — APPROVED** [review]
LGTM 👍

File: src/core/net/dnssd_server.hpp (hunk 1 above)
Comment (abtink) [reply src/core/net/dnssd_server.hpp:318]: We can consider using `InstanceLocatorInit` which provides an `Init()` function instead (this is what we use for other types which are used in array, e.g. `Neighbor`.

**Review by abtink — APPROVED** [review]
LGTM. Thanks.

One suggestion below (the current model is also fine as is in my view).

**Review by jwhui — APPROVED** [review]

File: src/core/net/dnssd_server.hpp [hunk 2]
```diff
@@ -403,6 +420,8 @@ class Server : public InstanceLocator, private NonCopyable
     otDnssdQuerySubscribeCallback   mQuerySubscribe;
     otDnssdQueryUnsubscribeCallback mQueryUnsubscribe;
     TimerMilli                      mTimer;
+
+    Counters mCounters;
```
Comment (jwhui) [inline src/core/net/dnssd_server.hpp:424]: Do we need to initialize `mCounters`?

Comment (superwhd) [reply src/core/net/dnssd_server.hpp:424]: Good catch! I just updated the constructor of `Server`.

Comment (abtink) [reply src/core/net/dnssd_server.hpp:424]: @superwhd I noticed you added `mCounters()` in `Server` constructor with the goal to have `mCounters `struct be initialized as zero.

While this should work in most cases (and I believe maybe even mandated by C++11 onwards), I suggest we do the clearing explicitly.

We can define
```
class Counters : public otDnssdCounters, public Clearable<Counter> 
{ 
};
``` 
Then we can use `Clear()` method (or add a constructor that does clear the counters explicitly).

The reason I am suggesting this is that I have seen some C++ toolchains (specially custom embedded tool chains) that may not do the zero-initialization properly (or may allow it to be disabled through certain config).

In general, it is safer to do this explicitly in OT code (it does not add any overhead and we we avoid weird behaviors/bugs due to OT being build on specific toolchains or under certain configs).

**Pull request merged**